### PR TITLE
adding kernels

### PR DESCRIPTION
**Pull request by tiandiao123**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/kernel/cuda_native/setup.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+                Path(this_dir)  / 'csrc'/'cutlass' / 'include',
+            ],
+        ),
+
+    ],
+    cmdclass={
+        'build_ext': BuildExtension
+})
+
+
```
Comment (ver217) [inline colossalai/kernel/cuda_native/setup.py:248]: Can you integrate this to colossalai's setup?

File: colossalai/kernel/cuda_native/csrc/attention_infer_kernels/flash_attn/src/block_info.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
+/******************************************************************************
```
Comment (ver217) [inline colossalai/kernel/cuda_native/csrc/attention_infer_kernels/flash_attn/src/block_info.h:1]: Are these files the same as origin flash attn?

File: colossalai/kernel/cuda_native/csrc/attention_infer_kernels/layernorm/layernorm_kernels.cu [hunk 3]
```diff
@@ -0,0 +1,62 @@
+/*This code from Vllm : https://github.com/vllm-project/vllm
+ *     with minor changes. */
```
Comment (ver217) [inline colossalai/kernel/cuda_native/csrc/attention_infer_kernels/layernorm/layernorm_kernels.cu:2]: We already had layernorm kernels. Is this more efficient?

File: colossalai/kernel/cuda_native/csrc/attention_infer_kernels/linear/linear_op.cpp [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+    // Delete copy and assignment constructors
+    CudaStream(const CudaStream&) = delete;
+    CudaStream& operator=(const CudaStream&) = delete;
+};
+
+
+
+
+void dense_layer_fp32_kernel(const float *in, const float *weight, float *out, const int M,
```
Comment (ver217) [inline colossalai/kernel/cuda_native/csrc/attention_infer_kernels/linear/linear_op.cpp:82]: As we discussed, this kernel is less efficient than pytorch's. Did we optimize this one? If not, we may not add this kernel to colossalai.

Comment (tiandiao123) [reply colossalai/kernel/cuda_native/csrc/attention_infer_kernels/linear/linear_op.cpp:82]: yes, I think we can delete this since linear implementation is not good.

File: colossalai/kernel/cuda_native/csrc/attention_infer_kernels/layernorm/layernorm_kernels.cu (hunk 3 above)
Comment (tiandiao123) [reply colossalai/kernel/cuda_native/csrc/attention_infer_kernels/layernorm/layernorm_kernels.cu:2]: yes, it is more efficient than apex layernorm

File: colossalai/kernel/cuda_native/csrc/attention_infer_kernels/flash_attn/src/block_info.h (hunk 2 above)
Comment (tiandiao123) [reply colossalai/kernel/cuda_native/csrc/attention_infer_kernels/flash_attn/src/block_info.h:1]: not exactly same, I only kept forward codes over here.

Comment (tiandiao123) [reply colossalai/kernel/cuda_native/csrc/attention_infer_kernels/flash_attn/src/block_info.h:1]: > Are these files the same as origin flash attn?

or we can directly use third-party (flash-attention-2) as our submodules?

File: colossalai/kernel/cuda_native/setup.py (hunk 1 above)
Comment (tiandiao123) [reply colossalai/kernel/cuda_native/setup.py:248]: yes!

File: colossalai/kernel/cuda_native/csrc/attention_infer_kernels/flash_attn/src/block_info.h (hunk 2 above)
Comment (ver217) [reply colossalai/kernel/cuda_native/csrc/attention_infer_kernels/flash_attn/src/block_info.h:1]: I think can we directly import and use it?

File: tests/test_kernels/cuda/test_rmsnorm.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    rms_norm(
+        out,
+        x,
+        weight,
+        variance_epsilon,
+    )
+    return out
+
+
+def test_rmsnorm():
```
Comment (ver217) [inline tests/test_kernels/cuda/test_rmsnorm.py:45]: You should add skip if here.

File: colossalai/kernel/cuda_native/csrc/attention_infer_kernels/flash_attn/src/block_info.h (hunk 2 above)
Comment (tiandiao123) [reply colossalai/kernel/cuda_native/csrc/attention_infer_kernels/flash_attn/src/block_info.h:1]: OK!

File: tests/test_kernels/cuda/test_rmsnorm.py (hunk 5 above)
Comment (tiandiao123) [reply tests/test_kernels/cuda/test_rmsnorm.py:47]: fixed it.

File: colossalai/kernel/cuda_native/csrc/attention_infer_kernels/rmsnorm/layernorm_kernels.cu [hunk 6]
```diff
@@ -0,0 +1,62 @@
+/*This code from Vllm : https://github.com/vllm-project/vllm
```
Comment (ver217) [inline colossalai/kernel/cuda_native/csrc/attention_infer_kernels/rmsnorm/layernorm_kernels.cu:1]: Should we add their license to ours?

File: colossalai/kernel/cuda_native/csrc/attention_infer_kernels/softmax/fused_softmax.cpp [hunk 7]
```diff
@@ -0,0 +1,46 @@
+/* coding=utf-8
+ * Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
```
Comment (ver217) [inline colossalai/kernel/cuda_native/csrc/attention_infer_kernels/softmax/fused_softmax.cpp:2]: Is it from apex? Can we use it directly?

File: op_builder/llama_infer_cuda.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+from packaging.version import parse, Version
+
+from setuptools import setup, find_packages
+import subprocess
+
+import torch
+from setuptools import setup
+from torch.utils.cpp_extension import BuildExtension, CUDAExtension, CUDA_HOME
+from torch.utils import cpp_extension
+
```
Comment (ver217) [inline op_builder/llama_infer_cuda.py:16]: Why don't you use the Builder class?

File: colossalai/kernel/cuda_native/csrc/attention_infer_kernels/softmax/fused_softmax.cpp (hunk 7 above)
Comment (tiandiao123) [reply colossalai/kernel/cuda_native/csrc/attention_infer_kernels/softmax/fused_softmax.cpp:2]: it is not from apex, it is from faster transformer

File: op_builder/llama_infer_cuda.py (hunk 8 above)
Comment (tiandiao123) [reply op_builder/llama_infer_cuda.py:16]: I can change to use build class.

File: colossalai/kernel/cuda_native/csrc/attention_infer_kernels/softmax/fused_softmax.cpp (hunk 7 above)
Comment (tiandiao123) [reply colossalai/kernel/cuda_native/csrc/attention_infer_kernels/softmax/fused_softmax.cpp:2]: currently, I saw there already has some scaled softmax in kernels, I can temporarily remove this. If I feel the original one is not useful, I can add it back.

File: colossalai/kernel/cuda_native/csrc/attention_infer_kernels/rmsnorm/layernorm_kernels.cu (hunk 6 above)
Comment (tiandiao123) [reply colossalai/kernel/cuda_native/csrc/attention_infer_kernels/rmsnorm/layernorm_kernels.cu:1]: I am not quite sure. We can remove this temporarily

Comment (ver217) [reply colossalai/kernel/cuda_native/csrc/attention_infer_kernels/rmsnorm/layernorm_kernels.cu]: NO. I mean we should add their license to ours (`ColossalAI/LICENSE`)

File: tests/test_kernels/cuda/test_rotary_embedding.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    )
+    ref_query = ref_query.view(num_tokens, num_heads * head_size)
+    ref_key = ref_key.view(num_tokens, num_heads * head_size)
+
+    # Compare the results.
+    assert torch.allclose(out_query, ref_query, atol=1e-3, rtol=1e-5)
+    assert torch.allclose(out_key, ref_key, atol=1e-3, rtol=1e-5)
+
+@pytest.mark.skipif(not HAS_INFER_CUDA, reason="You need to install llama supported cuda kernels to run this test")
+def test():
```
Comment (ver217) [inline tests/test_kernels/cuda/test_rotary_embedding.py:145]: It should have a meaningful name, like `test_xxx()`

**Review by ver217 — APPROVED** [review]
